rescan.go

```diff
@@ -399,10 +399,153 @@ filterHeaderWaitLoop:
 		})
 	}
 
+	// We'll need to keep track of whether we are current with the chain in
+	// order to properly recover from a re-org. We'll start by assuming that
+	// we are not current in order to catch up from the starting point to
+	// the tip of the chain.
+	current := false
+
+	// handleBlockConnected is a closure that handles a new block connected
+	// notification.
+	//
+	// TODO(wilmer): refactor this and handleBlockDisconnected into their
+	// own methods.
+	handleBlockConnected := func(header wire.BlockHeader) error {
+		// If we've somehow missed a header in the range, then we'll
+		// mark ourselves as not current so we can walk down the chain
+		// and notify the callers of blocks we may have missed.
+		//
+		// It's possible due to the nature of the current subscription
+		// system that we get a duplicate block. We'll catch this and
+		// continue forwards to avoid an unnecessary state transition
+		// back to the !current state.
+		if header.PrevBlock != curStamp.Hash &&
+			header.BlockHash() != curStamp.Hash {
+
+			current = false
+			return fmt.Errorf("out of order block %v: "+
+				"expected PrevBlock %v, got %v",
+				header.BlockHash(), curStamp.Hash,
+				header.PrevBlock)
+		}
+
+		// Do not process block until we have all filter headers. Don't
+		// worry, the block will get re-queued every time there is a new
+		// filter available. However, if it's a duplicate block
+		// notification, then we can re-process it without any issues.
+		if header.BlockHash() != curStamp.Hash &&
+			!s.hasFilterHeadersByHeight(uint32(curStamp.Height+1)) {
+
+			log.Warnf("Missing filter header for height=%v, "+
+				"skipping", curStamp.Height+1)
+
+			return nil
+		}
+
+		// As this could be a re-try, we'll ensure that we don't
+		// incorrectly increment our current time stamp.
+		if curStamp.Hash != header.BlockHash() {
+			curHeader = header
+			curStamp.Hash = header.BlockHash()
+			curStamp.Height++
+		}
+
+		log.Tracef("Rescan got block %d (%s)", curStamp.Height,
+			curStamp.Hash)
+
+		// We're only scanning if the header is beyond the horizon of
+		// our start time.
+		if !scanning {
+			scanning = ro.startTime.Before(
+				curHeader.Timestamp,
+			)
+		}
+
+		// If we're actually scanning and we have a non-empty watch
+		// list, then we'll attempt to fetch the filter from the
+		// network.
```

Comment:
This section doesn't look like it's prevented from executing if we're not scanning or have an empty watch list? If that's the case, it will get filters regardless of whether we're scanning or not. Since you're just moving this code, this could be a TODO to update the code and/or comment.

Reply:
Fixed.